File: companies/New Folder With Items/3916.number-of-zig-zag-arrays-iii.java
/*
 * @lc app=leetcode id=3916 lang=java
 *
 * [3916] Number of Zig Zag Arrays III
 *
 * Block matrix exponentiation on flat arrays; symmetric squaring when base is symmetric.
 */

// @lc code=start
class Solution {
    private static final int MOD = 1_000_000_007;

    public int zigZagArrays(int n, int l, int r) {
        int mod = MOD;
        int m = r - l + 1;

        if (n == 1) {
            return m % mod;
        }
        if (n == 2) {
            return (int) ((long) m * (m - 1) % mod);
        }

        int ex = n - 2;

        int[] ls = buildLFlat(m);
        int[] rs = buildRFlat(m);
        int[] lr = buildMinMatrixFlat(m, mod);
        int[] rl = buildMaxMatrixFlat(m, mod);

        int[] vd = new int[m];
        int[] vu = new int[m];
        for (int j = 0; j < m; j++) {
            vd[j] = j;
            vu[j] = m - 1 - j;
        }

        if (ex % 2 == 0) {
            int k = ex / 2;
            int[] plr = matPowSymmetricBase(lr, m, k, mod);
            int[] prl = matPowSymmetricBase(rl, m, k, mod);
            int[] a = vecMatFlat(vd, plr, m, mod);
            int[] b = vecMatFlat(vu, prl, m, mod);
            return (int) ((sum(a) + sum(b)) % mod);
        } else {
            int k = ex / 2;
            int[] plr = matPowSymmetricBase(lr, m, k, mod);
            int[] prl = matPowSymmetricBase(rl, m, k, mod);
            int[] tmpU = matMulFlat(prl, rs, m, mod);
            int[] tmpD = matMulFlat(plr, ls, m, mod);
            int[] a = vecMatFlat(vu, tmpU, m, mod);
            int[] b = vecMatFlat(vd, tmpD, m, mod);
            return (int) ((sum(a) + sum(b)) % mod);
        }
    }

    private static long sum(int[] a) {
        long s = 0;
        for (int v : a) {
            s += v;
        }
        return s;
    }

    private static int[] buildLFlat(int m) {
        int[] mat = new int[m * m];
        for (int i = 0; i < m; i++) {
            int base = i * m;
            for (int j = 0; j < i; j++) {
                mat[base + j] = 1;
            }
        }
        return mat;
    }

    private static int[] buildRFlat(int m) {
        int[] mat = new int[m * m];
        for (int i = 0; i < m; i++) {
            int base = i * m;
            for (int j = i + 1; j < m; j++) {
                mat[base + j] = 1;
            }
        }
        return mat;
    }

    private static int[] buildMinMatrixFlat(int m, int mod) {
        int[] mat = new int[m * m];
        for (int i = 0; i < m; i++) {
            int row = i * m;
            for (int j = 0; j < m; j++) {
                mat[row + j] = Math.min(i, j) % mod;
            }
        }
        return mat;
    }

    private static int[] buildMaxMatrixFlat(int m, int mod) {
        int[] mat = new int[m * m];
        for (int i = 0; i < m; i++) {
            int row = i * m;
            for (int j = 0; j < m; j++) {
                mat[row + j] = (m - 1 - Math.max(i, j)) % mod;
            }
        }
        return mat;
    }

    private static int[] matPowSymmetricBase(int[] matrix, int m, int p, int mod) {
        int[] r = new int[m * m];
        for (int i = 0; i < m; i++) {
            r[i * m + i] = 1;
        }
        int[] cur = matrix.clone();
        int pow = p;
        while (pow > 0) {
            if ((pow & 1) == 1) {
                r = matMulFlat(r, cur, m, mod);
            }
            cur = symmetricSquareFlat(cur, m, mod);
            pow >>= 1;
        }
        return r;
    }

    private static int[] symmetricSquareFlat(int[] m0, int m, int mod) {
        int[] c = new int[m * m];
        for (int i = 0; i < m; i++) {
            int ib = i * m;
            for (int j = i; j < m; j++) {
                int jb = j * m;
                long s = 0;
                for (int k = 0; k < m; k++) {
                    s += (long) m0[ib + k] * m0[jb + k];
                }
                int v = (int) (s % mod);
                c[ib + j] = v;
                if (i != j) {
                    c[j * m + i] = v;
                }
            }
        }
        return c;
    }

    private static int[] matMulFlat(int[] a, int[] b, int m, int mod) {
        int[] c = new int[m * m];
        for (int i = 0; i < m; i++) {
            int ib = i * m;
            for (int k = 0; k < m; k++) {
                int aik = a[ib + k];
                if (aik == 0) {
                    continue;
                }
                int kb = k * m;
                for (int j = 0; j < m; j++) {
                    c[ib + j] = (int) ((c[ib + j] + (long) aik * b[kb + j]) % mod);
                }
            }
        }
        return c;
    }

    private static int[] vecMatFlat(int[] v, int[] mat, int m, int mod) {
        int[] out = new int[m];
        for (int j = 0; j < m; j++) {
            long s = 0;
            for (int i = 0; i < m; i++) {
                int vi = v[i];
                if (vi != 0) {
                    s = (s + (long) vi * mat[i * m + j]) % mod;
                }
            }
            out[j] = (int) s;
        }
        return out;
    }
}
// @lc code=end

File: pt-1/3916.number-of-zig-zag-arrays-iii.go
package leetcode

//
// @lc app=leetcode id=3916 lang=golang
//
// [3916] Number of Zig Zag Arrays III
//
// Correctness (odd length exponent): v·T^(2k+1) is NOT vd·(LR)^(k+1)+vu·(RL)^(k+1).
// With T=[[0,L],[R,0]], one gets new D-block = vu·(RL)^k·R and new U-block = vd·(LR)^k·L.
//
// Performance (CP Python):
// - Block reduction: (2m)×(2m) → m×m matrices LR=L·R, RL=R·L (LR[i,j]=min(i,j),
//   RL[i,j]=m−1−max(i,j)).
// - Flat row-major int lists (lower overhead than list-of-lists).
// - Symmetric squaring: if M is symmetric, (M²)[i,j]=Σ_k M[i,k]M[j,k]; fill only
//   i≤j and mirror → ~2× fewer ops when computing cur·cur in binary exponentiation.
// - General multiply kept for nonsymmetric R·cur.
//
// NumPy is intentionally NOT used: modular integer matmul needs big integers or
// per-step `%`, and object-dtype loops are slower than tight Python int loops here.
//
// =============================================================================

// @lc code=start

const mod3916 int64 = 1000000007

// ZigZagArrays3916 returns the number of zig-zag arrays III modulo 1e9+7.
func ZigZagArrays3916(n int, l int, r int) int {
	m := r - l + 1
	if n == 1 {
		return int(int64(m) % mod3916)
	}
	if n == 2 {
		return int((int64(m) * int64(m-1)) % mod3916)
	}

	ex := n - 2

	Ls := buildLFlat3916(m)
	Rs := buildRFlat3916(m)
	LR := buildMinMatrixFlat3916(m)
	RL := buildMaxMatrixFlat3916(m)

	vd := make([]int64, m)
	vu := make([]int64, m)
	for j := 0; j < m; j++ {
		vd[j] = int64(j)
		vu[j] = int64(m - 1 - j)
	}

	var a, b []int64
	if ex%2 == 0 {
		k := ex / 2
		Plr := matPowSymmetricBase3916(LR, m, k)
		Prl := matPowSymmetricBase3916(RL, m, k)
		a = vecMatFlat3916(vd, Plr, m)
		b = vecMatFlat3916(vu, Prl, m)
	} else {
		k := ex / 2
		Plr := matPowSymmetricBase3916(LR, m, k)
		Prl := matPowSymmetricBase3916(RL, m, k)
		// T^(2k+1) has blocks (LR)^k·L on U<-D and (RL)^k·R on D<-U :
		//   new(D-part) = vu · (RL)^k · R ,   new(U-part) = vd · (LR)^k · L
		tmpU := matMulFlat3916(Prl, Rs, m)
		tmpD := matMulFlat3916(Plr, Ls, m)
		a = vecMatFlat3916(vu, tmpU, m)
		b = vecMatFlat3916(vd, tmpD, m)
	}

	sum := int64(0)
	for _, x := range a {
		sum += x
	}
	for _, x := range b {
		sum += x
	}
	return int(sum % mod3916)
}

func buildLFlat3916(m int) []int64 {
	// D_i → U_j: L[i,j]=1 iff j<i (strict lower 1s in value index).
	M := make([]int64, m*m)
	for i := 0; i < m; i++ {
		base := i * m
		for j := 0; j < i; j++ {
			M[base+j] = 1
		}
	}
	return M
}

func buildRFlat3916(m int) []int64 {
	// U_i → D_j: R[i,j]=1 iff j>i.
	M := make([]int64, m*m)
	for i := 0; i < m; i++ {
		base := i * m
		for j := i + 1; j < m; j++ {
			M[base+j] = 1
		}
	}
	return M
}

func buildMinMatrixFlat3916(m int) []int64 {
	M := make([]int64, m*m)
	for i := 0; i < m; i++ {
		row := i * m
		for j := 0; j < m; j++ {
			if i < j {
				M[row+j] = int64(i)
			} else {
				M[row+j] = int64(j)
			}
		}
	}
	return M
}

func buildMaxMatrixFlat3916(m int) []int64 {
	M := make([]int64, m*m)
	for i := 0; i < m; i++ {
		row := i * m
		for j := 0; j < m; j++ {
			mx := i
			if j > mx {
				mx = j
			}
			M[row+j] = int64(m - 1 - mx)
		}
	}
	return M
}

func matPowSymmetricBase3916(M []int64, m int, p int) []int64 {
	// Binary exponentiation; base matrix M is symmetric so use symmetric squaring.
	R := make([]int64, m*m)
	for i := 0; i < m; i++ {
		R[i*m+i] = 1
	}
	cur := make([]int64, len(M))
	copy(cur, M)
	for p > 0 {
		if (p & 1) == 1 {
			R = matMulFlat3916(R, cur, m)
		}
		cur = symmetricSquareFlat3916(cur, m)
		p >>= 1
	}
	return R
}

func symmetricSquareFlat3916(M []int64, m int) []int64 {
	// For symmetric M: C[i,j]=Σ_k M[i,k]M[j,k]; compute i≤j only.
	C := make([]int64, m*m)
	for i := 0; i < m; i++ {
		ib := i * m
		for j := i; j < m; j++ {
			jb := j * m
			var s int64
			for k := 0; k < m; k++ {
				s += (M[ib+k] * M[jb+k]) % mod3916
			}
			v := s % mod3916
			C[ib+j] = v
			if i != j {
				C[j*m+i] = v
			}
		}
	}
	return C
}

func matMulFlat3916(A []int64, B []int64, m int) []int64 {
	// ikj order (good cache); one `%` per output cell after summing all k.
	C := make([]int64, m*m)
	for i := 0; i < m; i++ {
		ib := i * m
		for k := 0; k < m; k++ {
			aik := A[ib+k]
			if aik == 0 {
				continue
			}
			kb := k * m
			for j := 0; j < m; j++ {
				C[ib+j] += aik * B[kb+j]
			}
		}
		for j := 0; j < m; j++ {
			C[ib+j] %= mod3916
		}
	}
	return C
}

func vecMatFlat3916(v []int64, M []int64, m int) []int64 {
	out := make([]int64, m)
	for j := 0; j < m; j++ {
		var s int64
		for i := 0; i < m; i++ {
			vi := v[i]
			if vi != 0 {
				s = (s + (vi*M[i*m+j])%mod3916) % mod3916
			}
		}
		out[j] = s
	}
	return out
}

// @lc code=end

